### ISSUE TITLE
Changes to LintRule

### ISSUE THREAD

**Issue by bartlomieju** [denoland/deno_lint#424]
- Add `LintRule::lint_program(&self, context: &mut Context, program: &swc_ecmascript::ast::Program)` in favor of `LintRule::lint_module` - prerequisite for #366

- Rules should implement `Visit` trait directly on the struct implementing `LintRule` instead of creating additional visitor for each rule

**Comment by RDambrosio016** [comment]
You will encounter some problems when it comes to the second point. I tried this with the original version of rslint (back when it used a visitor which was made with the same macro as swc's) but it was a huge hack since you get context for a rule when you run it, but you also need to access that context through each visit function. Therefore you have to either construct the rule with the context already (and perhaps have a lifetime for the context) or give each visit function a mut reference to the context, which makes for a pretty long function definition for each visit function

**Comment by bartlomieju** [comment]
@RDambrosio016 good point, I actually want to go other way. I took a closer look at [`declare_lint` from RSLint](https://github.com/RDambrosio016/RSLint/blob/0eb52aa41f8a9583423ec0b23c9ab34779617d75/crates/rslint_core/src/rule.rs#L226-L274) and I think it's pretty useful. 

For `deno_lint` I'm thinking something along the lines:
```rust
define_lint_rule! {
  name: BanTsComment,
  code: "ban-ts-comment",
  tags: ["recommended"],
  docs: r#"Disallows the use of Typescript directives without a comment.

  Typescript directives reduce the effectiveness of the compiler, something which should only be done in exceptional circumstances.  The reason why should be documented in a comment alongside the directive.
  ..."#,
  handler: (
    rule: &BanTsComment,
    context: &mut Context,
    program: &swc_ecmascript::ast::Program,
  ) {
    let mut violated_comment_spans = Vec::new();

    violated_comment_spans.extend(
      context.leading_comments.values().flatten().filter_map(|c| {
        if check_comment(c) {
          Some(c.span)
        } else {
          None
        }
      }),
    );
    violated_comment_spans.extend(
      context
        .trailing_comments
        .values()
        .flatten()
        .filter_map(|c| if check_comment(c) { Some(c.span) } else { None }),
    );

    for span in violated_comment_spans {
      rule.report(context, span);
    }
  },
}
```

which would expand to:

```rust
impl LintRule for BanTsComment {
  fn new() -> Box<Self> {
    Box::new(BanTsComment)
  }

  fn tags(&self) -> &[&'static str] {
    &["recommended"]
  }

  fn code(&self) -> &'static str {
    "ban-ts-comment"
  }

  fn handler(
    &self,
    context: &mut Context,
    _program: &swc_ecmascript::ast::Program,
  ) {
    let mut violated_comment_spans = Vec::new();

    violated_comment_spans.extend(
      context.leading_comments.values().flatten().filter_map(|c| {
        if check_comment(c) {
          Some(c.span)
        } else {
          None
        }
      }),
    );
    violated_comment_spans.extend(
      context
        .trailing_comments
        .values()
        .flatten()
        .filter_map(|c| if check_comment(c) { Some(c.span) } else { None }),
    );

    for span in violated_comment_spans {
      self.report(context, span);
    }
  }

  fn docs(&self) -> &'static str {
    r#"Disallows the use of Typescript directives without a comment.

Typescript directives reduce the effectiveness of the compiler, something which should only be done in exceptional circumstances.  The reason why should be documented in a comment alongside the directive.
..."#
  }
}
```

So that would actually mean that most of visitors stay the same, but declaration is less boilerplaty

**Comment by RDambrosio016** [comment]
I do not reccomend that for a couple of reasons:
1: you can capture docs from macro invocations, granted, it requires a bit of a cursed "recursive muncher" but it would allow you to make both rustdoc and user facing docs instead of only user facing docs.
2: you will get really bad autocompletion (or none at all most likely), at least for me, RA does not do autocompletion for macros when you take in `$($body:tt)*`

**Comment by bartlomieju** [comment]
Discussed with @RDambrosio016 offline. I think we could settle on:
```rust
define_lint_rule! {
  name: BanTsComment,
  code: "ban-ts-comment",
  tags: ["recommended"],
}
```

which would expand to:

```rust
pub struct BanTsComment;

impl LintRule for BanTsComment {
  fn new() -> Box<Self> {
    Box::new(BanTsComment)
  }

  fn tags(&self) -> &[&'static str] {
    &["recommended"]
  }

  fn code(&self) -> &'static str {
    "ban-ts-comment"
  }
}
```

Writing docs and handler would still require to write `impl BanTsComment { ... }` but that's fine, it's not that much boilerplate anymore.

**Comment by sanderhahn** [comment]
@bartlomieju was looking if i can this to work...

The `LintRule` trait has more methods besides `tags` and `code`, for example: `lint_program` and `docs`.
Do you want to embed them inside the `define_lint_rule!` block like this?

```rust
macro_rules! define_lint_rule {
    (
        name: $type:ident,
        code: $ident:ident,
        tags: $expr:expr,
        $( $body:tt )*
    ) => {
        pub struct $type;
        impl LintRule for $type {
            fn new() -> Box<Self> {
              Box::new($type)
            }

            fn tags(&self) -> &'static [&'static str] {
              &$expr
            }

            fn code(&self) -> &'static str {
              $ident
            }

            $( $body )*
        }
    }
}
```

**Comment by RDambrosio016** [comment]
That still suffers from the second issue i talked about

**Comment by magurotuna** [comment]
IMHO that `define_lint_rule!` macro doesn't look really useful to me, because our current way of implementing `LintRule` trait is not so redundant. Plus as @RDambrosio016 says, using the token tree type can lead to bad developer experience

**Comment by bartlomieju** [comment]
> The LintRule trait has more methods besides tags and code, for example: lint_program and docs.
Do you want to embed them inside the define_lint_rule! block like this?

@sanderhahn no I only meant to define struct, code and tags using macro as per @RDambrosio016's suggestion that defining method in macro is error prone. 

@magurotuna yeah, but there's a lot of `'static` in typings and it's really cumbersome to type it all out and it would also be a bit easier to do refactors if we were to deduplicate that code.

**Comment by RDambrosio016** [comment]
Could also do it the way rslint does and have a rule trait with the metadata and a CstRule trait with the actual implementation

**Comment by sanderhahn** [comment]
Sorry new to this codebase and Rust and macro's in general. Maybe you could create a `#[lint_rule(code="ban-types",tags="recommended",docs="docs/ban_types.md")]` on the rule structs?

**Comment by RDambrosio016** [comment]
Yes but:
- it requires another proc macro crate
- it still does not fix the issue of needing the implementation to be in the macro invocation

The core issue is that rust doesnt allow you to partially impl a trait then finish impl'ing it somewhere else, so you cannot impl the code, tags, and docs methods and then impl the lint method. Which is why i chose to go with a rule trait then a CstRule trait, that way:
- rules arent constrained to only run on a single file and you dont need hacks (looking at you eslint-plugin-import).
- you can add the metadata and the implementation separately (which is why rslint's `declare_lint` is a thing)

And finally, yes it is possible to capture doc comments into a string using declarative macros, [RSLint does it](https://github.com/RDambrosio016/RSLint/blob/master/crates/rslint_core/src/rule.rs#L351) but it requires a bit of a complex macro helper.

**Comment by bartlomieju** [comment]
After second thought I will close this issue now and maybe revisit in the future - we got more important stuff to tend to than adding helper macros. Thanks to everyone for opinions 👍
